### PR TITLE
Detect custom OpenAPI files and start stripe-mock from test suite

### PR DESCRIPTION
**Pull request by mickjermsurawong-stripe**
Check for custom open api files to run stripe-mock process from test suite.
This is the same work already implemented in other libs
https://github.com/stripe/stripe-dotnet/pull/1445
https://github.com/stripe/stripe-ruby/pull/715

I've verified that with the custom files detected, the build would pass in travis [here](https://github.com/stripe/stripe-java/compare/mickjermsurawong/stripe-mock-fixture...mickjermsurawong/stripe-mock-fixture-with-specs)

r? @ob-stripe 
cc @stripe/api-libraries

### CODE REVIEW

File: src/test/java/com/stripe/BaseStripeTest.java [hunk 1]
```diff
@@ -83,11 +104,11 @@ public static void checkStripeMock() throws Exception {
   }
 
   /**
-   * Activates stripe-mock by overriding the API host and putting a test key
-   * into the environment.
+   * Activates usage of stripe-mock by overriding the API host and putting a test key
+   * into the environment. This is required independent of how strip-mock is started.
```
[the remark this answers is not in the document]
Comment (mickjermsurawong-stripe) [reply src/test/java/com/stripe/BaseStripeTest.java]: fixed!

File: src/test/java/com/stripe/StripeMockProcess.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            String.valueOf(port),
+            "-spec",
+            getPathSpec(),
+            "-fixtures",
+            getPathFixture()
+        );
+
+    // stripe-mock output/error will be the same as this current test parent process
+    processBuilder.redirectOutput(ProcessBuilder.Redirect.INHERIT);
+    processBuilder.redirectError(ProcessBuilder.Redirect.INHERIT);
```
[the remark this answers is not in the document]
Comment (mickjermsurawong-stripe) [reply src/test/java/com/stripe/StripeMockProcess.java]: Reverted to default, set to `Redirect.Pipe`
https://docs.oracle.com/javase/7/docs/api/java/lang/ProcessBuilder.Redirect.html#PIPE

File: src/test/java/com/stripe/StripeMockProcess.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      ));
+      System.exit(1);
+    }
+
+    pid = getProcessId(process);
+
+    if (isAlive(process)) {
+      System.out.println(String.format("Started stripe-mock, PID = %s", pid));
+      System.out.println("Finding port bound to stripe-mock...");
+      port = detectBoundPort(process);
```
Comment (mickjermsurawong-stripe) [inline src/test/java/com/stripe/StripeMockProcess.java]: This would block if there's no process output. 
I could potentially make this non-blocking as a thread, and continually ping within the timeout.
Given the approach of output string pattern matching, we already assume the predictable pattern of our internal tooling, maybe it's not worthwhile to implement the time-out pattern catching when there's no output.
I also put this block after checking that stripe-mock is alive, to reduce chance of hanging when somehow the process exits.

Alternatively, I thought about using `lsof -Pan -p <stripe mock PID> -i` and get the response. This can make our port finding independent of stripe-mock implementation, but the downside is perhaps how we have variance to lsof string output which makes pattern matching more fragile. Mine looks something like this.
```
COMMAND     PID             USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
stripe-mo 20288 mickjermsurawong    3u  IPv6 0x4dxxxxx      0t0  TCP *:51516 (LISTEN)
```

Comment (mickjermsurawong-stripe) [reply src/test/java/com/stripe/StripeMockProcess.java]: I believe `readPartial` in ruby's implementation will have the same problem on "It blocks only if ios has no data immediately available." [doc](https://ruby-doc.org/core-2.2.2/IO.html#method-i-readpartial)

**Comment by mickjermsurawong-stripe** [comment]
thank you ob, I implemented a similar pattern for finding port to that in Ruby! ptal @ob-stripe

**Comment by mickjermsurawong-stripe** [comment]
hmm cibot doesn't reassign. let me try again:
r? @ob-stripe

File: src/test/java/com/stripe/StripeMockProcess.java (hunk 3 above)
Comment (ob-stripe) [reply src/test/java/com/stripe/StripeMockProcess.java]: Nice, thanks for digging into this Mick. I think we can just leave it at this. There's really no reason for the stripe-mock process to be alive and not output anything, and if that ever happens we can just look into it.

File: src/test/java/com/stripe/StripeMockProcess.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            getPathFixture()
+        );
+
+    try {
+      process = processBuilder.start();
+    } catch (IOException e) {
+      System.out.println(String.format(
+          "Error while starting stripe-mock, port = %d, fixtures = %s and %s, "
+              + "error message = %s",
+          port, getPathSpec(), getPathFixture(), e.getMessage()
```
Comment (ob-stripe) [inline src/test/java/com/stripe/StripeMockProcess.java]: Can you update this message to remove the port (since it will be `null` now that stripe-mock chooses the port itself)?

Comment (mickjermsurawong-stripe) [reply src/test/java/com/stripe/StripeMockProcess.java]: whoops my bad. fixed!

**Comment by mickjermsurawong-stripe** [comment]
Addressed the PR feedback. Thank you for the thorough review!
ptal @ob-stripe
